library/core/src/alloc/layout.rs

```diff
@@ -382,8 +391,10 @@ impl Layout {
     #[unstable(feature = "alloc_layout_extra", issue = "55724")]
     #[inline]
     pub fn repeat_packed(&self, n: usize) -> Result<Self, LayoutError> {
-        let size = self.size().checked_mul(n).ok_or(LayoutError)?;
-        Layout::from_size_align(size, self.align())
+        // Size manipulation is done in isize space to avoid overflowing isize.
+        let n = isize::try_from(n).map_err(|_| LayoutError)?;
```

Comment:
Is this check correct?  It looks like it would fail on `Layout::new::<()>().repeat_packed(usize::MAX)`, but I think that should be fine, since repeating a ZST wouldn't overflow `isize` *bytes*.

Reply:
You're correct that it incorrectly fails here.

I'm not sure how better to write this function, though; this could either be written as a `usize` `checked_mul` and an additional `< isize::MAX` check,

Reply:
I don't know; it would probably depend greatly on what information is available at the callsites.  (Whether the `self.size` or the `n` is more likely to be constant.)

Hmm, this might be too cute, but what about:
```rust
let n = (self.size << 1).checked_mul(n).ok_or(LayoutError)? >> 1;
```
along with a whole bunch of tests, of course.

(The shift can't overflow because `self.size` always has the high bit zero, so the unsigned overflow stops at the right place, and the shift back down at the end just shifts out a low-bit zero to fix the scale.)

Reply:
Godbolt comparison: https://rust.godbolt.org/z/a3sb3v3sK

Observation putting this together: we don't have to check `isize:MAX` here; `from_size_align` will of course now do that for us, meaning the implementation *does not change* from `HEAD`, since it was using the safe `from_size_align` already.

The size check when nothing is known is fairly clearly the worst option of the correct ones (3 jumps vs 1). Strangely enough, it looks like doing the check that way is potentially *better* than the `isize::checked_mul` (1 jump vs 2); based on staring at Godbolt it looks like this is because with the `usize` mul we get an overflow check error return and handle `from_size_align` with `cmov`s, whereas with the `isize` mul we need to insert a check and bail for an input high bit first, then do the same thing.

I think reverting this to HEAD beats out the bit tricks by obviousness of implementation alone, given the minimal output difference, but it also looks to me like the bit twiddling impl ends up executing all the same assembly instructions but with an additional two to shl and shr, such that an overflow occurs sooner. If anything, this nano-optimizes for the overflowing case, which we expect to be the less likely case anyway.

Reply:
I like the plan of just using `from_size_align`.  As you say, the obvious correctness is hugely valuable, and if nothing's clearly better right now, we can always μoptimize later. 👍 

(I still plan to make a #95361-like change for the size, which might help optimize `Layout` without needing to drop to `unsafe`, by moving more of the checks to the type system.)